File: BigDataSpark/scripts/reports_clickhouse.py
from pyspark.sql import SparkSession
from pyspark.sql import functions as F

# Инициализация Spark сессии
spark = SparkSession.builder.appName("ReportsClickHouse").getOrCreate()

pg_url = "jdbc:postgresql://postgres:5432/postgres"
pg_props = {
    "user": "postgres",
    "password": "postgres",
    "driver": "org.postgresql.Driver"
}

ch_url = "jdbc:clickhouse://clickhouse:8123/default"
ch_driver = "com.clickhouse.jdbc.ClickHouseDriver"

fact = spark.read.jdbc(url=pg_url, table="fact_sales", properties=pg_props)
dim_p = spark.read.jdbc(url=pg_url, table="products", properties=pg_props)
dim_c = spark.read.jdbc(url=pg_url, table="customers", properties=pg_props)
dim_st = spark.read.jdbc(url=pg_url, table="stores", properties=pg_props)
dim_sup = spark.read.jdbc(url=pg_url, table="suppliers", properties=pg_props)

# 1. Витрина продаж по продуктам
top10_products = (
    fact.groupBy("product_id")
    .agg(
        F.sum("quantity").alias("total_quantity"),
        F.sum("total_price").alias("total_revenue")
    )
    .join(dim_p, "product_id")
    .select("product_id", "name", "category", "total_quantity", "total_revenue")
    .orderBy(F.desc("total_quantity"))
    .limit(10)
)

top10_products.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "top10_products") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

revenue_by_category = (
    fact.join(dim_p, "product_id")
    .groupBy("category")
    .agg(F.sum("total_price").alias("total_revenue"))
    .orderBy(F.desc("total_revenue"))
)

revenue_by_category.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "revenue_by_category") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

product_ratings = dim_p.select("product_id", "name", "category", "rating", "reviews")

product_ratings.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "product_ratings") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

# 2. Витрина продаж по клиентам
top10_customers = (
    fact.groupBy("customer_id")
    .agg(F.sum("total_price").alias("total_spent"))
    .join(dim_c, "customer_id")
    .select("customer_id", "first_name", "last_name", "country", "total_spent")
    .orderBy(F.desc("total_spent"))
    .limit(10)
)

top10_customers.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "top10_customers") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

customers_by_country = (
    dim_c.groupBy("country")
    .agg(F.countDistinct("customer_id").alias("num_customers"))
    .orderBy(F.desc("num_customers"))
)

customers_by_country.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "customers_by_country") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

avg_check_by_customer = (
    fact.groupBy("customer_id")
    .agg((F.sum("total_price") / F.count("quantity")).alias("avg_check"))
    .join(dim_c, "customer_id")
    .select("customer_id", "avg_check")
)

avg_check_by_customer.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "avg_check_by_customer") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

# 3. Витрина продаж по времени
monthly_trends = (
    fact.withColumn("year", F.year("sale_date"))
    .withColumn("month", F.month("sale_date"))
    .groupBy("year", "month")
    .agg(
        F.sum("total_price").alias("revenue"),
        F.sum("quantity").alias("quantity")
    )
    .withColumn("avg_order_size", F.col("revenue") / F.col("quantity"))
    .orderBy("year", "month")
)

monthly_trends.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "monthly_trends") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

yearly_trends = (
    fact.withColumn("year", F.year("sale_date"))
    .groupBy("year")
    .agg(
        F.sum("total_price").alias("revenue"),
        F.sum("quantity").alias("quantity")
    )
    .orderBy("year")
)

yearly_trends.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "yearly_trends") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

# 4. Витрина продаж по магазинам
top5_stores = (
    fact.groupBy("store_id")
    .agg(F.sum("total_price").alias("revenue"))
    .join(dim_st, "store_id")
    .select("name", "city", "country", "revenue")
    .orderBy(F.desc("revenue"))
    .limit(5)
)

top5_stores.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "top5_stores") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

sales_by_city_country = (
    fact.join(dim_st, "store_id")
    .groupBy("city", "country")
    .agg(
        F.sum("total_price").alias("revenue"),
        F.sum("quantity").alias("quantity")
    )
    .orderBy(F.desc("revenue"))
)

sales_by_city_country.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "sales_by_city_country") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

avg_check_by_store = (
    fact.groupBy("store_id")
    .agg((F.sum("total_price") / F.count("quantity")).alias("avg_check"))
    .join(dim_st, "store_id")
    .select("name", "avg_check")
)

avg_check_by_store.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "avg_check_by_store") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

# 5. Витрина продаж по поставщикам
top5_suppliers = (
    fact.groupBy("supplier_id")
    .agg(F.sum("total_price").alias("revenue"))
    .join(dim_sup, "supplier_id")
    .select("name", "city", "country", "revenue")
    .orderBy(F.desc("revenue"))
    .limit(5)
)

top5_suppliers.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "top5_suppliers") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

avg_price_by_supplier = (
    fact.groupBy("supplier_id")
    .agg(F.avg(F.col("total_price") / F.col("quantity")).alias("avg_price"))
    .join(dim_sup, "supplier_id")
    .select("name", "avg_price")
)

avg_price_by_supplier.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "avg_price_by_supplier") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

sales_by_supplier_country = (
    fact.join(dim_sup, "supplier_id")
    .groupBy("country")
    .agg(
        F.sum("total_price").alias("revenue"),
        F.sum("quantity").alias("quantity")
    )
    .orderBy(F.desc("revenue"))
)

sales_by_supplier_country.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "sales_by_supplier_country") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

# 6. Витрина качества продукции
highest_rated = (
    dim_p.orderBy(F.desc("rating"))
    .limit(10)
    .select("product_id", "name", "rating")
)

highest_rated.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "highest_rated_products") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

lowest_rated = (
    dim_p.orderBy(F.asc("rating"))
    .limit(10)
    .select("product_id", "name", "rating")
)

lowest_rated.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "lowest_rated_products") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

rating_sales = (
    fact.join(dim_p, "product_id")
    .groupBy("product_id", "name")
    .agg(
        F.avg("rating").alias("avg_rating"),
        F.sum("quantity").alias("total_quantity")
    )
)

corr_value = rating_sales.stat.corr("avg_rating", "total_quantity")
corr_df = spark.createDataFrame([(corr_value,)], ["rating_sales_correlation"])

corr_df.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "rating_sales_correlation") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

top_reviewed = (
    dim_p.orderBy(F.desc("reviews"))
    .limit(10)
    .select("product_id", "name", "reviews")
)

top_reviewed.write.format("jdbc").mode("overwrite") \
    .option("url", ch_url) \
    .option("dbtable", "top_reviewed_products") \
    .option("driver", ch_driver) \
    .option("createTableOptions", "ENGINE = Log") \
    .save()

spark.stop()

File: BigDataSpark/src/main/java/ru/lunidep/ReportsClickHouse.java
package ru.lunidep;

import org.apache.spark.sql.*;
import org.apache.spark.sql.catalyst.expressions.GenericRow;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.Metadata;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import static org.apache.spark.sql.functions.*;

public class ReportsClickHouse {
    public static void main(String[] args) {
        SparkSession spark = SparkSession.builder()
                .appName("ReportsClickHouse")
                .getOrCreate();

        String pgUrl = "jdbc:postgresql://postgres:5432/postgres";
        java.util.Properties pgProps = new java.util.Properties();
        pgProps.setProperty("user", "postgres");
        pgProps.setProperty("password", "postgres");
        pgProps.setProperty("driver", "org.postgresql.Driver");

        String chUrl = "jdbc:clickhouse://clickhouse:8123/default";

        Dataset<Row> fact = spark.read().jdbc(pgUrl, "fact_sales", pgProps);
        Dataset<Row> dimP = spark.read().jdbc(pgUrl, "products", pgProps);
        Dataset<Row> dimC = spark.read().jdbc(pgUrl, "customers", pgProps);
        Dataset<Row> dimSt = spark.read().jdbc(pgUrl, "stores", pgProps);
        Dataset<Row> dimSup = spark.read().jdbc(pgUrl, "suppliers", pgProps);

        Dataset<Row> top10Products = fact.groupBy("product_id")
                .agg(sum("quantity").alias("total_quantity"),
                        sum("total_price").alias("total_revenue"))
                .join(dimP, "product_id")
                .select("product_id", "name", "category", "total_quantity", "total_revenue")
                .orderBy(col("total_quantity").desc())
                .limit(10);

        writeToClickHouse(top10Products, chUrl, "top10_products");

        Dataset<Row> revenueByCategory = fact.join(dimP, "product_id")
                .groupBy("category")
                .agg(sum("total_price").alias("total_revenue"))
                .orderBy(col("total_revenue").desc());

        writeToClickHouse(revenueByCategory, chUrl, "revenue_by_category");

        Dataset<Row> productRatings = dimP.select("product_id", "name", "category", "rating", "reviews");
        writeToClickHouse(productRatings, chUrl, "product_ratings");

        Dataset<Row> top10Customers = fact.groupBy("customer_id")
                .agg(sum("total_price").alias("total_spent"))
                .join(dimC, "customer_id")
                .select("customer_id", "first_name", "last_name", "country", "total_spent")
                .orderBy(col("total_spent").desc())
                .limit(10);

        writeToClickHouse(top10Customers, chUrl, "top10_customers");

        Dataset<Row> customersByCountry = dimC.groupBy("country")
                .agg(countDistinct("customer_id").alias("num_customers"))
                .orderBy(col("num_customers").desc());

        writeToClickHouse(customersByCountry, chUrl, "customers_by_country");

        Dataset<Row> avgCheckByCustomer = fact.groupBy("customer_id")
                .agg(sum("total_price").divide(count("quantity")).alias("avg_check"))
                .join(dimC, "customer_id")
                .select("customer_id", "avg_check");

        writeToClickHouse(avgCheckByCustomer, chUrl, "avg_check_by_customer");

        Dataset<Row> monthlyTrends = fact.withColumn("year", year(col("sale_date")))
                .withColumn("month", month(col("sale_date")))
                .groupBy("year", "month")
                .agg(sum("total_price").alias("revenue"),
                        sum("quantity").alias("quantity"))
                .withColumn("avg_order_size", col("revenue").divide(col("quantity")))
                .orderBy("year", "month");

        writeToClickHouse(monthlyTrends, chUrl, "monthly_trends");

        Dataset<Row> yearlyTrends = fact.withColumn("year", year(col("sale_date")))
                .groupBy("year")
                .agg(sum("total_price").alias("revenue"),
                        sum("quantity").alias("quantity"))
                .orderBy("year");

        writeToClickHouse(yearlyTrends, chUrl, "yearly_trends");

        Dataset<Row> top5Stores = fact.groupBy("store_id")
                .agg(sum("total_price").alias("revenue"))
                .join(dimSt, "store_id")
                .select("name", "city", "country", "revenue")
                .orderBy(col("revenue").desc())
                .limit(5);

        writeToClickHouse(top5Stores, chUrl, "top5_stores");

        Dataset<Row> salesByCityCountry = fact.join(dimSt, "store_id")
                .groupBy("city", "country")
                .agg(sum("total_price").alias("revenue"),
                        sum("quantity").alias("quantity"))
                .orderBy(col("revenue").desc());

        writeToClickHouse(salesByCityCountry, chUrl, "sales_by_city_country");

        Dataset<Row> avgCheckByStore = fact.groupBy("store_id")
                .agg(sum("total_price").divide(count("quantity")).alias("avg_check"))
                .join(dimSt, "store_id")
                .select("name", "avg_check");

        writeToClickHouse(avgCheckByStore, chUrl, "avg_check_by_store");

        Dataset<Row> top5Suppliers = fact.groupBy("supplier_id")
                .agg(sum("total_price").alias("revenue"))
                .join(dimSup, "supplier_id")
                .select("name", "city", "country", "revenue")
                .orderBy(col("revenue").desc())
                .limit(5);

        writeToClickHouse(top5Suppliers, chUrl, "top5_suppliers");

        Dataset<Row> avgPriceBySupplier = fact.groupBy("supplier_id")
                .agg(avg(col("total_price").divide(col("quantity"))).alias("avg_price"))
                .join(dimSup, "supplier_id")
                .select("name", "avg_price");

        writeToClickHouse(avgPriceBySupplier, chUrl, "avg_price_by_supplier");

        Dataset<Row> salesBySupplierCountry = fact.join(dimSup, "supplier_id")
                .groupBy("country")
                .agg(sum("total_price").alias("revenue"),
                        sum("quantity").alias("quantity"))
                .orderBy(col("revenue").desc());

        writeToClickHouse(salesBySupplierCountry, chUrl, "sales_by_supplier_country");

        Dataset<Row> highestRated = dimP.orderBy(col("rating").desc())
                .limit(10)
                .select("product_id", "name", "rating");

        writeToClickHouse(highestRated, chUrl, "highest_rated_products");

        Dataset<Row> lowestRated = dimP.orderBy(col("rating").asc())
                .limit(10)
                .select("product_id", "name", "rating");

        writeToClickHouse(lowestRated, chUrl, "lowest_rated_products");

        Dataset<Row> ratingSales = fact.join(dimP, "product_id")
                .groupBy("product_id", "name")
                .agg(avg("rating").alias("avg_rating"),
                        sum("quantity").alias("total_quantity"));

        double corrValue = ratingSales.stat().corr("avg_rating", "total_quantity");
        Dataset<Row> corrDf = spark.createDataFrame(
                java.util.Arrays.asList(
                        new GenericRow(new Object[]{corrValue})
                ),
                new StructType(new StructField[]{
                        new StructField("rating_sales_correlation", DataTypes.DoubleType, false, Metadata.empty())
                })
        );

        writeToClickHouse(corrDf, chUrl, "rating_sales_correlation");

        Dataset<Row> topReviewed = dimP.orderBy(col("reviews").desc())
                .limit(10)
                .select("product_id", "name", "reviews");

        writeToClickHouse(topReviewed, chUrl, "top_reviewed_products");

        spark.stop();
    }

    private static void writeToClickHouse(Dataset<Row> df, String url, String table) {
        df.write()
                .format("jdbc")
                .mode(SaveMode.Overwrite)
                .option("url", url)
                .option("dbtable", table)
                .option("driver", "com.clickhouse.jdbc.ClickHouseDriver")
                .option("createTableOptions", "ENGINE = Log")
                .save();
    }
}